src/rgw/rgw_op.cc

```diff
@@ -2666,9 +2733,60 @@ void RGWPutObj::execute()
     goto done;
   }
 
+  /* handle x-amz-copy-source */
+
+  if (copy_source) {
+    bucket = s->info.env->get("HTTP_X_AMZ_COPY_SOURCE");
```

Comment:
@jmunhoz should do this prior to execute(), probably in get_params(). As it is, it's a security issue, we don't check here permissions of the source object, which needs to happen.